lib/proxy/log_buffer/buffer.rb

```diff
@@ -0,0 +1,110 @@
+require 'date'
+require 'proxy/log_buffer/ring_buffer'
+
+# Adopted from Celluloid library (ring_buffer.rb).
+# Copyright (c) 2011-2014 Tony Arcieri. Distributed under the MIT License.
+# https://github.com/celluloid/celluloid/blob/0-16-stable/lib/celluloid/logging/ring_buffer.rb
+module Proxy::LogBuffer
+
+  LogRecord = Struct.new(:timestamp, :level, :message, :backtrace) do
+    def to_h
+      h = {}
+      self.class.members.each{|m| h[m.to_sym] = self[m]}
+      h[:level] = case h[:level]
+                  when ::Logger::Severity::INFO
+                    :INFO
+                  when ::Logger::Severity::WARN
+                    :WARN
+                  when ::Logger::Severity::ERROR
+                    :ERROR
+                  when ::Logger::Severity::FATAL
+                    :FATAL
+                  when ::Logger::Severity::DEBUG
+                    :DEBUG
+                  else
+                    :UNKNOWN
+                  end
+      h.delete(:backtrace) unless h[:backtrace]
+      h
+    end
+  end
+
+  class Buffer
+    def self.instance
+      @@buffer ||= Buffer.new
+    end
+
+    def initialize
+      @mutex = Mutex.new
+      @buffers_set_up = false
+      @failed_modules = {}
+    end
+
+    # late initialization
+    def setup_buffers(size = nil, size_tail = nil, level = nil, level_tail = nil)
```

Comment:
Why delay initialization? Doesn't look like anything can raise an exception here, other than requiring of proxy/settings (shouldn't need that, it's one of the first things to load).